File: Binary Search/binarySearch.js

//O(log n) time complexity

// iterative binary search
function binarySearch(target, array){
    let beg = 0;
    let end = array.length -1;
    let mid = Math.floor((beg+end)/2);
    
    while (array[mid] !== target && beg <= end && end >= beg){
        if (array[mid] > target) end = mid - 1;
        else beg = mid + 1;
        mid = Math.floor((beg+end)/2);
    }
    array[mid] === target ? console.log(mid) : false;
}

let nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20];
binarySearch(19, nums);

File: Binary Search/binarySearch.py

# iterative binary search
def binarySearch(target, array):
    beg = 0
    end = len(array) - 1
    mid = 0

    while beg <= end:
        mid = (end + beg)//2
        if array[mid] < target:
            beg = mid + 1
        elif array[mid] > target:
            end = mid - 1
        else:
            return mid
    return -1

# error handling for input
while True:    
    try:
        target = int(input('Select target (1-10):'))
        if target > 10 or target < 1:
            print('Error: out of range value')
        else: break
    except:
        print('Error: not a valid integer')
        continue

nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]    
result = binarySearch(target, nums)
print('Target', target, 'was found at index', result)

    

